internal/monitor/instance.go

```diff
@@ -0,0 +1,94 @@
+// SPDX-FileCopyrightText: Copyright (c) 2023, CIQ, Inc. All rights reserved
+// SPDX-License-Identifier: Apache-2.0
+package monitor
+
+import (
+	"bytes"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/jasonyangshadow/apptheus/internal/cgroup"
+	"github.com/jasonyangshadow/apptheus/internal/cgroup/parser"
+	"github.com/jasonyangshadow/apptheus/internal/push"
+	"github.com/jasonyangshadow/apptheus/storage"
+)
+
+type Instance struct {
+	*cgroup.CGroup
+	ticker *time.Ticker
+
+	ErrCh chan error
+	Done  chan struct{}
+}
+
+func New(ticker *time.Ticker) *Instance {
+	ins := &Instance{}
+	ins.ticker = ticker
+	ins.ErrCh = make(chan error, 1)
+	ins.Done = make(chan struct{}, 1)
+	return ins
+}
+
+func (i *Instance) Start(container *parser.ContainerInfo, ms storage.MetricStore, logger log.Logger) {
+	c, err := cgroup.NewCGroup(container.ID)
+	if err != nil {
+		level.Error(logger).Log("msg", "while validating cgroup info", "err", err, "container id", container.ID)
+		i.ErrCh <- err
+		return
+	}
+	i.CGroup = c
+
+	err = i.Apply(int(container.Pid))
+	if err != nil {
+		level.Error(logger).Log("msg", "while adding proc to cgroup info", "err", err, "container id", container.ID)
+		i.ErrCh <- err
+		return
+	}
+
+	defer i.Destroy()
+
+	var buffer bytes.Buffer
+	labels := make(map[string]string)
+	labels["job"] = container.ID
+
+	for {
+		for range i.ticker.C {
```

Comment:
also the ticker is never stopped, could be a resource leak issue for the long runs

Reply:
I added a defer ticker.Stop() in the beginning of the function.